auto_inertial_params/proposal.md

```diff
@@ -0,0 +1,287 @@
+
+  /// \brief SDF element pointer to <auto_inertia_params> tag.
+  /// This can be used to access custom params for the
+  /// Inertia Caluclator
+  public: sdf::ElementPtr inertiaCalculatorParams{nullptr};
+};
+```
+
+This class would act as an interface between `libsdformat` and the custom calculator to bridge the data like density, properties of the mesh and user-defined calculator params given through the SDF. A signature for the Custom Calculator function is provided through an alias:
+
+```C++
+using CustomInertiaCalculator = std::function<std::optional<gz::math::Inertiald>(sdf::Errors &, const sdf::CustomInertiaCalcProperties &)>;
+```
+
+Functions to get and register a custom inertia calculator is provided through the `sdf::ParserConfig`. Using all these additions, the `Mesh::CalculateInertial()` function would be implmented as follows:
+
+```C++
+  /// \brief Calculate and return the Mass Matrix values for the Mesh
+  /// \param[in] density Density of the mesh in kg/m^3
+  /// \param[in] _autoInertiaParams ElementPtr to
+  /// <auto_inertia_params> element
+  /// \param[in] _config Parser Configuration
+  /// \param[out] _errors A vector of Errors object. Each object
+  /// would contain an error code and an error message.
+  /// \return A std::optional with gz::math::Inertiald object or std::nullopt
+  public: std::optional<gz::math::Inertiald> CalculateInertial(double _density, const sdf::ElementPtr _autoInertiaParams, const ParserConfig &_config, sdf::Errors &_errors);
+
+  //////////////////////////////////////////////////
+  std::optional<gz::math::Inertiald> Mesh::CalculateInertial(double _density,
+      const sdf::ElementPtr _autoInertiaParams, const ParserConfig &_config,
+      sdf::Errors &_errors)
+  {
+    if (this->dataPtr->filePath.empty())
+    {
+      _errors.push_back({
+        sdf::ErrorCode::WARNING,
+        "File Path for the mesh was empty. Could not calculate inertia"});
+      return std::nullopt;
+    }
+
+    const auto &customCalculator = _config.CustomInertiaCalc();
+
+    sdf::CustomInertiaCalcProperties calcInterface = CustomInertiaCalcProperties(
+      _density, *this, _autoInertiaParams);
+
+    return customCalculator(_errors, calcInterface);
+  }
+```
+
+### Configuring the `CalculateInertial()` function behavior
+
+`CalculateInertial()` functions are also added to the `sdf::Root`, `sdf::World`, `sdf::Model`, `sdf::Link`, `sdf::Collision` and then `sdf::Geometry` classes. Calling the `CalculateInertial()` function of any class, in turn recursively calls the `CalculateInertial()` for each element below it down the chain. For eg, generally a user might call the `Root::CalculateInertial()` which would in turn call `World::CalculateInertials()` for all the worlds which in turn calls the `Model::CaluclateInertials()` for all the models in the world and so on until the final `CalculateInertial()` for the respective geometry type is called where the chain ends.
```

Comment:
```suggestion
`CalculateInertial()` functions are also added to the `sdf::Root`, `sdf::World`, `sdf::Model`, `sdf::Link`, `sdf::Collision` and then `sdf::Geometry` classes. Calling the `CalculateInertial()` function of any class, recursively calls the `CalculateInertial()` for each element below it down the chain. For eg, generally a user adds a callback to `Root::CalculateInertial()` which calls `World::CalculateInertials()` for all the worlds, which in turn calls the `Model::CaluclateInertials()` for all the models in the world and so on until the final `CalculateInertial()` for the respective geometry type is called where the chain ends.
```

Reply:
Corrected in [8b902d8](https://github.com/gazebosim/sdf_tutorials/pull/92/commits/8b902d835468ab130a8d2a832dedfd444e2013b0)